### PR TITLE
add `simulate` parameter to `TURN`

### PR DESCRIPTION
**Pull request by ArnoldSmith86**
This PR adds `readonly: true` to `TURN` so that you can check which seats will become active without actually changing anything.

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: https://test.virtualtabletop.io/PR-1222/pr-test (or any other room on that server)

File: client/js/widgets/widget.js [hunk 1]
```diff
@@ -1496,11 +1496,17 @@ export class Widget extends StateManaged {
           const turn = indexList[mod(nextTurnIndex, indexList.length)];
 
           collections[a.collection] = [];
-          //saves turn into all seats and creates output collection with turn seats
-          for(const w of c) {
-            await w.set('turn', w.get('index') == turn);
-            if(w.get('turn') && w.get('player'))
-              collections[a.collection].push(w);
+
+          if(readonly) {
```
Comment (RaphaelAlvez) [inline client/js/widgets/widget.js:1500]: probably this should be readOnly

also didn't test but why doesn't this need to be a.readOnly like every property in a routine?

**Review by RaphaelAlvez — CHANGES_REQUESTED** [review]

**Comment by RaphaelAlvez** [comment]
btww I really like the concept just a minor change needed in my opinion

**Comment by RaphaelAlvez** [comment]
I made the changes and tested them. I added my test room to the test server. I took the liberty to open this PR to reviews

File: client/js/widgets/widget.js [hunk 2]
```diff
@@ -1496,14 +1496,21 @@ export class Widget extends StateManaged {
...
+          if(!a.readOnly) {
+            //saves turn into all seats and creates output collection with turn seats
+            for(const w of c) {
+              await w.set('turn', w.get('index') == turn);
+              if(w.get('turn') && w.get('player'))
+                collections[a.collection].push(w);
+            }
+            jeLoggingRoutineOperationSummary(`changed turn of seats from ${previousTurn} to ${turn} - active seats: ${JSON.stringify(indexList)}`);
+          } else {
+            collections[a.collection] = c.filter(w=>w.get('index')==turn && w.get('player'));
```
Comment (rogerl50) [inline client/js/widgets/widget.js:1509]: Should the `if` statement in the other branch be changed to this? They appear to do the same thing.

**Comment by 96LawDawg** [comment]
This is very straightforward and works as expected except for `"turncycle":"random"`.  If we are going to implement it, we should probably figure out a way to make that happen.

**Comment by RaphaelAlvez** [comment]
> This is very straightforward and works as expected except for `"turncycle":"random"`.  If we are going to implement it, we should probably figure out a way to make that happen.

Good call. Two ways to see this:
- simulate the TURN with readOnly and then actually do the TURN to just set the turn using the new preset turn variable.
- add internal code that makes the second TURN not actually random. Just get the previous result from the simulated TURN

File: client/js/widgets/widget.js (hunk 2 above)
Comment (RaphaelAlvez) [reply client/js/widgets/widget.js]: done

File: client/js/widgets/widget.js (hunk 1 above)
Comment (RaphaelAlvez) [reply client/js/widgets/widget.js]: done

File: client/js/jsonedit.js [hunk 3]
```diff
@@ -647,7 +647,7 @@ function jeAddCommands() {
   jeAddRoutineOperationCommands('SORT', { key: 'value', reverse: false, locales: null, options: null, holder: null, collection: 'DEFAULT' });
   jeAddRoutineOperationCommands('SHUFFLE', { holder: null, collection: 'DEFAULT' });
   jeAddRoutineOperationCommands('TIMER', { value: 0, seconds: 0, mode: 'toggle', timer: null, collection: 'DEFAULT' });
-  jeAddRoutineOperationCommands('TURN', { turn: 1, turnCycle: 'forward', source: 'all', collection: 'TURN' });
+  jeAddRoutineOperationCommands('TURN', { turn: 1, turnCycle: 'forward', source: 'all', collection: 'TURN', readOnly: false });
```
Comment (rogerl50) [inline client/js/jsonedit.js:650]: I'm late to the party, but I'd prefer something like `dryRun` to `readOnly`. You aren't really just reading properties, you are doing computations.

File: client/js/widgets/widget.js [hunk 4]
```diff
@@ -1497,16 +1497,22 @@ export class Widget extends StateManaged {
           if(typeof nextTurnIndex != 'number' || !isFinite(nextTurnIndex))
             nextTurnIndex = 0;
           const turn = indexList[mod(nextTurnIndex, indexList.length)];
+          variables.TURN = turn;
```
Comment (rogerl50) [inline client/js/widgets/widget.js:1500]: I don't think this works for `random` (and in fact in the pr-test room it doesn't). I *think* you are intending to have `variables.TURN` contain the new saved turn. But on this line, you have to set it if `readOnly` is true, and clear it if `readOnly` is false. Then, in the `a.turnCycle == 'random'` branch, you need to test `variables.TURN` and either just use it or call `random`.

File: client/js/widgets/widget.js [hunk 5]
```diff
@@ -1497,16 +1497,22 @@ export class Widget extends StateManaged {
...
+              await w.set('turn', w.get('index') == turn);
+            }
+            jeLoggingRoutineOperationSummary(`changed turn of seats from ${previousTurn} to ${turn} - active seats: ${JSON.stringify(indexList)}`);
+          } else {
+            jeLoggingRoutineOperationSummary(`SIMULATED turn of seats: from ${previousTurn} to ${turn} - active seats: ${JSON.stringify(indexList)}`);
           }
 
-          jeLoggingRoutineOperationSummary(`changed turn of seats from ${previousTurn} to ${turn} - active seats: ${JSON.stringify(indexList)}`);
+          collections[a.collection] = c.filter(w=>w.get('index')==turn && w.get('player'));
+
```
Comment (rogerl50) [inline client/js/widgets/widget.js:1515]: Can't this stay inside the `!a.readOnly` branch? If it's readonly, none of these things have changed, so the collection will remain unchanged.

**Review by rogerl50 — CHANGES_REQUESTED** [review]

File: client/js/widgets/widget.js (hunk 4 above)
Comment (RaphaelAlvez) [reply client/js/widgets/widget.js:1500]: I still think the way it is at the moment is good enough as described in my previous coment

File: client/js/widgets/widget.js (hunk 5 above)
Comment (RaphaelAlvez) [reply client/js/widgets/widget.js:1515]: @ArnoldSmith86?

File: client/js/widgets/widget.js (hunk 4 above)
Comment (rogerl50) [reply client/js/widgets/widget.js:1500]: Wait, I don't think you can leave it as is. It just doesn't work for `random` (that is, if you run `readOnly` with `random`, and then run it for real, you will likely get different results. I think I proposed a fix above that would work.

**Comment by 96LawDawg** [comment]
Through testing in the current room, random and readOnly always produce a turn index of 1.

**Comment by rogerl50** [comment]
> Through testing in the current room, random and readOnly always produce a turn index of 1.

Am I doing something wrong? I added labels above each button, and simulating followed by the real thing doesn't produce the same result.

**Comment by 96LawDawg** [comment]
> Am I doing something wrong? I added labels above each button, and simulating followed by the real thing doesn't produce the same result.

No.  I was.

**Comment by 96LawDawg** [comment]
I need this PR.  So I spent time fixing it.  It's been abandoned for months, so I'm hoping it can get a little attention soonish because now it is working.

**Comment by 96LawDawg** [comment]
So there is one possible problem with the approach I took.  I hate to even bring it up because this will probably cause more controversy, but here goes.  If you use random in simulate mode, it stores the random seat in `variables.TURN`.  That will stay there until a non-simulated TURN of any type is executed.  So it is possible that the code does a random simulation that results in seat2.  Then seat2 leaves.  Then the actual "random" TURN is done and selects seat2 which nobody is in.

I don't have a solution for that.  I don't think there is a solution for that.  I thought about putting something in the seat code to reset `variables.TURN` if the player is leaving and has been assigned the next random iteration.  But that would mean there would be a difference between what the JSON is expecting for a random result and what would actually happen.

**Comment by ArnoldSmith86** [comment]
Question is how this is supposed to work in general. Different calls to `TURN` could come from different widgets and with different parameters.

For everything but random, it should be straightforward, I think? Since it's deterministic, it can just focus on its own parameters and doesn't have to remember anything.

My guess is that for random, we return the simulated value for the next `TURN` _only if_ all other parameters of the `TURN` are identical to the simulation call _and_ if that value is still a valid option.

Thoughts?

**Comment by 96LawDawg** [comment]
> My guess is that for random, we return the simulated value for the next `TURN` _only if_ all other parameters of the `TURN` are identical to the simulation call _and_ if that value is still a valid option.
> 

That sounds like a good solution, I guess.  But I'm not sure what is better from a game designer point of view.  Should I get the seat that I expected even if it isn't valid.  Or do I get a null result and have to do it again.  I guess null and your solution is the safest.  Not sure how to implement that in the code.

**Comment by ArnoldSmith86** [comment]
Hmm, maybe we just relegate it to the game designer then? Return the simulated value and the designer has to save it and use it in the next call to `TURN`?

**Comment by 96LawDawg** [comment]
> Hmm, maybe we just relegate it to the game designer then? Return the simulated value and the designer has to save it and use it in the next call to `TURN`?

With my solution, the simulated value is saved.  You can do the simulation, do other stuff, and then later in a separate widget in a separate routine, get the random seat.  It is saved at least until the room is reloaded.  When the room is reloaded, the value is lost.

**Comment by ArnoldSmith86** [comment]
Hmm, that makes me like handing it off to the designer even more. Routines shouldn't behave differently if you reload the room between them.

**Comment by 96LawDawg** [comment]
> Routines shouldn't behave differently if you reload the room between them.

I think the behavior of simulated random TURN in this case is consistent with other routines, right?  If I GET a value from a widget, after reloading that room that value is no longer available to me.

There is probably a way to write `variables.TURN` in a more persistent way (like you can always get `playerColor`, but I don't think that makes things better.  Whatever helps everybody get to yes.  Because I don't really have a use case in mind for simulated random.  Just thought we should support it and now with do with a slight limitation that we can document in the wiki.

**Comment by robartsd** [comment]
I agree that simulating a random turn should not be expected to match the next random turn, they should remain completely independent, game designer needs to save the value to reuse if desired.

**Comment by rogerl50** [comment]
> I need this PR. So I spent time fixing it. It's been abandoned for months, so I'm hoping it can get a little attention soonish because now it is working.

Can you say specifically what the use case for this bit of functionality is?

**Comment by rogerl50** [comment]
Agree with just leaving this up to the game designer by returning the value. Even ignoring the `random` problem, there seem to be too many unfortunate cases.

**Comment by 96LawDawg** [comment]
> Can you say specifically what the use case for this bit of functionality is?

I thought I needed this to handle a feature in the trick-taking cards room I'm working on.  Turns out I did't, but this is what I was thinking:  Players are able to pre-select a card in their hand to play automatically.  I was going to use the turn simulator and check and see if that player would be the next player.  But I resolved it just by playing it automatically when the turn moves to that seat (seems obvious in retrospect).

But I could still see a use case.  Let's say it is Player 2's turn.  In the rules of the game, the next player and only the next can take some action to counteract what the player in front of him did.  Without changing the actual turn indicator, the automation has no way of knowing who the next player is.  Maybe it is seat 3, but if that seat is empty, then it might be seat 4, or seat 1.  The only way around that is to build your own array of seats and modulo through them (like what we did in poker before there were seats).  This simulate function makes the process easy.

**Comment by RaphaelAlvez** [comment]
> I agree that simulating a random turn should not be expected to match the next random turn, they should remain completely independent, game designer needs to save the value to reuse if desired.

Yeah @96LawDawg That was what I said before. The way it was implemented before you could simulate, store the resulting variable and use it later by using turnCycle position.

A problem I just realised we need a turnCycle "index" because the output TURN returns the index and not the position. So this aproach does not work

On top of that I don't like that your variables are hidden but I like the idea of a variable for that. @ArnoldSmith86 is there a way to have a variable that is permanent in the room. I was thinking on having currentTurn and nextTurn. current turn is the index of the seat with turn true and the next trun would be a settable global variable that the simulate TRUN would use. But If a game designer could read it or even change it.

So @96LawDawg I think you did pick a mistake in this approach but I think you tackled the wrong problem. this would be solved if we added a turnCycle index.

So the logic would go as this:
1. simulate TURN
2. save resulting variable into a property of any widget
3. when you finally want to execute the TURN use `trunCycle: index` and `turn: ${PROPERTY whatever}`

**Comment by 96LawDawg** [comment]
I don't think `turnCycle:index` is necessary.  On my local machine, I reverted everything I did.  Now with the plan (if I understand this right to just give the designer a variable to work with), this is pretty simple.

Do this to simulate:

```
{
      "func": "TURN",
      "readOnly": true,
      "turnCycle": "random"
    },
    {
      "func": "LABEL",
      "label": "turnStorage",
      "value": "${TURN}"
    }
```

And this to actually execute:
   ```
   {
      "func": "TURN",
      "turnCycle": "position",
      "turn": "${PROPERTY text OF turnStorage}"
    }
```

I haven't changed anything in the PR yet because I want to make sure this is right.  Need to change `readOnly` parameter to `simulate` again.  Anything else?

**Comment by 96LawDawg** [comment]
> To persist the value after the routine exits it must be written to the shared state so that it is shared across clients (currently routines can only write to state by setting widget properties, but we could add something stored in _meta); otherwise the data would be gone if the client refreshed for any reason.

Seems like overkill to me.  I think we have a simple solution that works now.  Just write it to a property like so many other automation things require.

Speaking of that, should we change the name of the returned variable to something other than `TURN`?  We already have a `TURN` function and a `turn` parameter.  Maybe we should call it `simTurn' or `nextSeat` or `nextTurn`?  But not make it persistent as Raphael was asking about.

**Comment by robartsd** [comment]
> Speaking of that, should we change the name of the returned variable to something other than `TURN`?

No, we should add a variable parameter to be consistent with other functions. It should default to TURN or TURN_INDEX, but the game designer should be able to specify whatever name they want.

**Comment by RaphaelAlvez** [comment]
We need to:
- revert @96LawDawg 
- add index input

**Comment by ArnoldSmith86** [comment]
Superseded by #1724.

**Pull request closed without merging**